### PR TITLE
Metrics crd changes to support authentication and POST requests

### LINKED ISSUE #195: [Feature Request] Enhance Metrics CRD to support for different types of metric database authentication
**Is your feature request related to a problem? Please describe.**
This issue deals with the enhancement needed for metric resources in order to support different types of database authentication.

This issue is part of the larger Iter8 issue documented [here](https://github.com/iter8-tools/iter8/issues/588).

**Why is the feature useful for Iter8 users?**
Please see [here](https://github.com/iter8-tools/iter8/issues/588).

**Describe the solution you'd like**
Introduce `authType` in metrics CRD, with valid values for `authType` being `basic`, `bearer`, and `apikey`.

**Describe alternatives you've considered**
Raw header based mechanisms supported as of now... not easy to use at all, and highly unintuitive from end-user perspective.

**Does this issue require a design doc/discussion? If there is a link to the design document/discussions, please provide it below.**
Yes, please see [here](https://github.com/iter8-tools/iter8/issues/588).

**How will this feature be tested?**
- [ ] When a metric is created with the `authType` field, the metric is accepted and the `authType` field is preserved (in other words, this field is part of metric CRD).
- [ ] Permissible values for `authType` are `basic`, `bearer`, and `apikey`.
- [ ] When a metric is created without `authType`, it is accepted and the `authType` field is not defaulted (in other words, no authentication will be attempted for this metric).

**How will this feature be documented?**
- [ ] etc3 developer documentation
- [ ] Please see [here](https://github.com/iter8-tools/iter8/issues/588) for end-user documentation requirements.

**Additional context**
This is a CRD change. Ensure that this is accomplished in conformance with [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md).

**References:**
1. http://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml
2. https://docs.python-requests.org/en/master/user/authentication/
3. https://www.loginradius.com/blog/async/everything-you-want-to-know-about-authorization-headers/
4. https://en.wikipedia.org/wiki/Basic_access_authentication
5. https://swagger.io/docs/specification/authentication/

### LINKED ISSUE #196: [Feature Request] Enhance Metrics CRD to support for GET and POST methods  for querying metric DBs
**Is your feature request related to a problem? Please describe.**
This enhancement needed for metric resources in order to support GET and POST methods while querying metric DBs.

This issue is part of the larger Iter8 issue documented [here](https://github.com/iter8-tools/iter8/issues/586).

**Why is the feature useful for Iter8 users?**
Please see [here](https://github.com/iter8-tools/iter8/issues/586).

**Describe the solution you'd like**
Introduce `method` in metrics CRD, with valid values for method being GET and POST.

**Describe alternatives you've considered**
There do not appear to be other alternatives in order to support POST queries.

**Does this issue require a design doc/discussion? If there is a link to the design document/discussions, please provide it below.**
Please see [here](https://github.com/iter8-tools/iter8/issues/586).

**How will this feature be tested?**
- [ ] When a metric is created with the `method` field, the metric is accepted and the `method` field is preserved (in other words, this field is part of metric CRD).
- [ ] Permissible values for `method` are `GET`, and `POST`.
- [ ] When a metric is created without `method`, it is accepted and the `method` field is defaulted to `GET`.

**How will this feature be documented?**
- [ ] etc3 developer documentation
- [ ] Please see [here](https://github.com/iter8-tools/iter8/issues/586) for end-user documentation requirements

**Additional Context**
This is a CRD change. Ensure that this is accomplished in conformance with [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md).

### PR DESCRIPTION
**Pull request by sriumcp**
Resolves #195 
Resolves #196 
Tests for all the behaviors identified in #195 and #196

### CODE REVIEW

File: api/v2alpha2/metric_types.go [hunk 1]
```diff
@@ -14,91 +14,132 @@ See the License for the specific language governing permissions and
...
 
 package v2alpha2
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// MetricType identifies the type of the metric
-// +kubebuilder:validation:Enum=counter;gauge
+// MetricType identifies the type of the metric.
```
Comment (sriumcp) [inline api/v2alpha2/metric_types.go:30]: Some fields will have a list of allowed values (enumerations). These values will be strings, and they will be in CamelCase, with an initial uppercase letter. Examples: ClusterFirst, Pending, ClientIP

File: api/v2alpha2/metric_types.go [hunk 2]
```diff
@@ -14,91 +14,132 @@ See the License for the specific language governing permissions and
...
+	CounterMetricType MetricType = "Counter"
 
-	// GaugeMetricType is an enhancement of Prometheus gauge metric type
-	GaugeMetricType MetricType = "gauge"
+	// GaugeMetricType is an enhancement of Prometheus Gauge metric type
+	GaugeMetricType MetricType = "Gauge"
 )
 
-// MetricSpec defines the desired state of Metric
+// AuthType identifies the type of authentication used in the HTTP request
```
Comment (sriumcp) [inline api/v2alpha2/metric_types.go:42]: Some fields will have a list of allowed values (enumerations). These values will be strings, and they will be in CamelCase, with an initial uppercase letter. Examples: ClusterFirst, Pending, ClientIP

When an acronym or initialism each letter in the acronym should be uppercase, such as with ClientIP or TCPDelay

File: api/v2alpha2/metric_types.go [hunk 3]
```diff
@@ -14,91 +14,132 @@ See the License for the specific language governing permissions and
...
+	// BearerAuthType corresponds to authentication with bearer token
+	BearerAuthType AuthType = "Bearer"
+
+	// APIKeyAuthType corresponds to authentication with API keys
+	APIKeyAuthType AuthType = "APIKey"
+)
+
+// MethodType identifies the HTTP request method (aka verb) used in the HTTP request
+// +kubebuilder:validation:Enum=GET;POST
+type MethodType string
```
Comment (sriumcp) [inline api/v2alpha2/metric_types.go:59]: When a proper name or the name of a command-line executable is used as a constant the proper name should be represented in consistent casing - examples: systemd, iptables, IPVS, cgroupfs, Docker (as a generic concept), docker (as the command-line executable). If a proper name is used which has mixed capitalization like eBPF that should be preserved in a longer constant such as eBPFDelegation.

File: api/v2alpha2/metric_types.go [hunk 4]
```diff
@@ -14,91 +14,132 @@ See the License for the specific language governing permissions and
...
+// MetricSpec defines the attributes of the Metric
 type MetricSpec struct {
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// Params are key/value pairs used to construct a REST query to the metrics backend
+	// Params are key/value pairs corresponding to HTTP request parameters
+	// Value may be templated, in which Iter8 will attempt to substitute placeholders in the template at query time using version information.
 	// +optional
-	Params *[]NamedValue `json:"params,omitempty" yaml:"params,omitempty"`
+	Params []NamedValue `json:"params,omitempty" yaml:"params,omitempty"`
```
Comment (sriumcp) [inline api/v2alpha2/metric_types.go:76]: Optional fields have the following properties:

* They have the +optional comment tag in Go.
* They are a pointer type in the Go definition (e.g. AwesomeFlag *SomeFlag) or have a built-in nil value (e.g. maps and slices).

Note the keyword "or" above. Slices have built in `nil` values so we will not use pointer type for them (FYI, slices and maps are internally represented using pointers in Go, so passing them by value is still highly efficient).

File: api/v2alpha2/metric_types.go [hunk 5]
```diff
@@ -14,91 +14,132 @@ See the License for the specific language governing permissions and
...
+	// Units of the metric. Used for informational purposes.
 	// +optional
 	Units *string `json:"units,omitempty" yaml:"units,omitempty"`
 
 	// Type of the metric
-	// +kubebuilder:default:="gauge"
+	// +kubebuilder:default:="Gauge"
 	// +optional
-	Type MetricType `json:"type" yaml:"type"`
+	Type *MetricType `json:"type,omitempty" yaml:"type,omitempty"`
```
Comment (sriumcp) [inline api/v2alpha2/metric_types.go:89]: Optional fields have the following properties:

They have the +optional comment tag in Go.
They are a pointer type in the Go definition (e.g. AwesomeFlag *SomeFlag) or have a built-in nil value (e.g. maps and slices).
The API server should allow POSTing and PUTing a resource with this field unset.

File: api/v2alpha2/metric_types.go [hunk 6]
```diff
@@ -14,91 +14,132 @@ See the License for the specific language governing permissions and
...
+	// The value of the SampleSize metric denotes the number of data points over which this metric is computed.
+	// This field is relevant only when Type == Gauge
 	// +optional
 	SampleSize *string `json:"sampleSize,omitempty" yaml:"sampleSize,omitempty"`
 
-	// Provider identifies the metric backend including its authentication properties and its unmarshaller
-	// +kubebuilder:validation:MinLength:=1
+	// AuthType is the type of authentication used in the HTTP request
+	// +optional
+	AuthType *AuthType `json:"authType,omitempty" yaml:"authType,omitempty"`
```
Comment (sriumcp) [inline api/v2alpha2/metric_types.go:99]: Optional fields have the following properties:

They have the +optional comment tag in Go.
They are a pointer type in the Go definition (e.g. AwesomeFlag *SomeFlag) or have a built-in nil value (e.g. maps and slices).
The API server should allow POSTing and PUTing a resource with this field unset.

File: api/v2alpha2/metric_types.go [hunk 7]
```diff
@@ -14,91 +14,132 @@ See the License for the specific language governing permissions and
...
+
+	// Method is the HTTP method used in the HTTP request
+	// +kubebuilder:default:="GET"
+	// +optional
+	Method *MethodType `json:"method,omitempty" yaml:"method,omitempty"`
```
Comment (sriumcp) [inline api/v2alpha2/metric_types.go:104]: Optional fields have the following properties:

They have the +optional comment tag in Go.
They are a pointer type in the Go definition (e.g. AwesomeFlag *SomeFlag) or have a built-in nil value (e.g. maps and slices).
The API server should allow POSTing and PUTing a resource with this field unset.

File: api/v2alpha2/metric_types.go [hunk 8]
```diff
@@ -14,91 +14,132 @@ See the License for the specific language governing permissions and
...
+
+	// Body is the string used to construct the (json) body of the HTTP request
+	// Body may be templated, in which Iter8 will attempt to substitute placeholders in the template at query time using version information.
+	// This field is relevant only when Method == POST
 	// +optional
-	Provider string `json:"provider" yaml:"provider"`
+	Body *string `json:"body,omitempty" yaml:"body,omitempty"`
```
Comment (sriumcp) [inline api/v2alpha2/metric_types.go:110]: Optional fields have the following properties:

They have the +optional comment tag in Go.
They are a pointer type in the Go definition (e.g. AwesomeFlag *SomeFlag) or have a built-in nil value (e.g. maps and slices).
The API server should allow POSTing and PUTing a resource with this field unset.

File: api/v2alpha2/metric_types.go [hunk 9]
```diff
@@ -14,91 +14,132 @@ See the License for the specific language governing permissions and
...
 
-	// JQExpression defines how the value of the metric should be extracted from the metric value
-	// returned by a query to the backend metrics server
-	// +kubebuilder:validation:MinLength:=1
+	// Provider identifies the type of metric database. Used for informational purposes.
 	// +optional
+	Provider *string `json:"provider,omitempty" yaml:"provider,omitempty"`
```
Comment (sriumcp) [inline api/v2alpha2/metric_types.go:114]: Optional fields have the following properties:

They have the +optional comment tag in Go.
They are a pointer type in the Go definition (e.g. AwesomeFlag *SomeFlag) or have a built-in nil value (e.g. maps and slices).
The API server should allow POSTing and PUTing a resource with this field unset.

File: api/v2alpha2/metric_types.go [hunk 10]
```diff
@@ -14,91 +14,132 @@ See the License for the specific language governing permissions and
...
-	// HeaderTemplates are templates for headers that should be passed to the metrics backend.
-	// Typically these are authentication headers. Any fields of the form '$name' are treated as
-	// variables whose value is looked up in the secret referred to by spec.secretRef. If the
-	// secret does not contain the variable as a key, the full string "$name" is used.
+	// HeaderTemplates are key/value pairs corresponding to HTTP request headers and their values.
+	// Value may be templated, in which Iter8 will attempt to substitute placeholders in the template at query time using Secret.
+	// Placeholder substitution will be attempted only when Secret != nil.
 	// +optional
-	HeaderTemplates *[]NamedValue `json:"headerTemplates,omitempty" yaml:"headerTemplates,omitempty"`
+	HeaderTemplates []NamedValue `json:"headerTemplates,omitempty" yaml:"headerTemplates,omitempty"`
```
Comment (sriumcp) [inline api/v2alpha2/metric_types.go:130]: Optional fields have the following properties:

They have the +optional comment tag in Go.
They are a pointer type in the Go definition (e.g. AwesomeFlag *SomeFlag) or have a built-in nil value (e.g. maps and slices).

Note the keyword "or" above. Slices have built in nil values so we will not use pointer type for them (FYI, slices and maps are internally represented using pointers in Go, so passing them by value is still highly efficient).

**Comment by kalantar** [comment]
@sriumcp changes to CamelCase for acceptable values has implications for a number of other projects. Issues should be created to update these.  These include (not sure I got them all). I don't think any require delaying this pull request but they should be addressed before we forget all the changes.

- iter8-kfserving - test yaml (this can probably wait if we are changing this stack)
- handler - test yaml, copy of crds
- iter8ctl - test yaml
- metrics mock - test yaml, main.go (can it use the constant value instead of hardcoding)
- iter8-install - metrics definitions

**Comment by sriumcp** [comment]
1. Iter8-kfserving repo is unsupported (not just deprecated). The idea is absorb it into the Iter8 repo shortly.
2. handler issue is here: https://github.com/iter8-tools/handler/issues/26
3. iter8ctl issue is here: https://github.com/iter8-tools/iter8ctl/issues/25
4. Metrics-mock types are not tied to etc3, though is a good idea to camel case them also for consistency. We can discuss this separately since metrics-mock will now undergo several other changes to accommodate authentication/POST APIs. Specifically, not changing metrics-mock at this point will not affect any existing tests, since it does not have an etc3 dependency.
5. iter8-install issue is here https://github.com/iter8-tools/iter8-install
6. Iter8 issue (not mentioned above, but is critical in terms of end-user documentation) is here: https://github.com/iter8-tools/iter8/issues/616

**Comment by sriumcp** [comment]
@kalantar All of the issues have now been created and identified.

**Review by kalantar — APPROVED** [review]

**Pull request merged**